subsys/pm/policy.c

```diff
@@ -136,6 +137,12 @@ const struct pm_state_info *pm_policy_next_state(uint8_t cpu, int32_t ticks)
 	uint8_t num_cpu_states;
 	const struct pm_state_info *cpu_states;
 
+#if CONFIG_PM_DEVICE
+	if (pm_device_is_any_busy()) {
```

Comment:
Also, take a look at the state locking API. It's the device driver code that needs to prevent system sleep if it's busy doing something.

Reply:
Out Intel ISH platform requires all devices are not busy to enter into system low power. Add state locking in every device driver may not be a suitable solution. Is it possible to add a Kconfig option in pm that enable such property, for example config NEED_ALL_DEVICES_UNBUSY, if this config is y then runs this pm_device_is_any_busy() checking in policy.c?

Reply:
here is a bad case:
- a driver sets device busy, starts one transaction, then waits the completion semaphore with a long timeout or even FOREVER.
- Zephyr runs to idle thread and gets a long idle time, so PM policy returns a deep sleep state.
- PM suspends all other devices, then SoC goes into very low PM state, maybe suspend to disk.
- SoC detects there's device interrupt pending, so wakes up immediately.
- do all resume works.
- device interrupt gets served.
in such a case, low-power states should be blocked. Any driver should avoid it. It's like a common thing. It seems not good for all drivers to call pm state lock APIs.

Reply:
If your device is waiting on something, and needs to prevent deep sleep, simply lock deep sleep states while waiting.

Reply:
It seems to me that most drivers have similar working model:
- set register to trigger a transaction.
- wait complete semaphore which will be given in ISR.
- complete the transaction.
all those drivers should call PM state lock APIs to lock all low power states after step 1. as a common driver, I've afraid they should call PM state lock API multi times to lock PM_STATE_SUSPEND_TO_IDLE, PM_STATE_STANDBY, PM_STATE_SUSPEND_TO_RAM and PM_STATE_SUSPEND_TO_DISK, then call the API multi times to unlock them after transaction done.
this seems not good at all.

Reply:
yeah, we know the policy code is common. That's why Leifu asked if it's possible to add the change wrapped with a Kconfig.
wondering how Linux kernel handle that... downloaded Linux kernel but I cannot find the place until now.
thanks @gmarull

Reply:
current solution doesn't solve the problems I mentioned.

Reply:
It is fine for me having this option wrapped around a Kconfig option if that is a common needed for devices in a platform. If it is particular to one device, I do stay with the state lock solution, but if that is common thing for the platform, I think this solution is reasonable, it less error prone.

Reply:
all ISH devices except two/three on AON domain need it.

Reply:
If ISH has more specific cases in term of pm_state selection, You can always define platform-specific policy `CONFIG_PM_POLICY_CUSTOM`.

Reply:
> If ISH has more specific cases in term of pm_state selection, You can always define platform-specific policy `CONFIG_PM_POLICY_CUSTOM`.

agree, why not use a custom policy instead of change the default policy. PM_NEED_ALL_DEVICES_IDLE seems to be very specialized and soon enough we will be wondering what is the usecase behind it without too much context and details in the documentation. See https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/subsys/pm/power_mgmt/src/main.c#L222C29-L222C49 for an example of a custom policy.